RCORE-2011: New parameter 'progress_estimate' for sync progress notification

kiburtse opened the pull request:
## What, How & Why?

* This PR introduces new progress estimate (value from 0.0 to 1.0) in existing progress notification's api
* New field is added to SyncSession::ProgressNotifierCallback which is also used in c-api and in AsycOpenTask
* Newly exposed field is supposed to provide current progress of synchronization for upload/download until completion
* Estimate for download with flexible sync configuration is provided by the server-side implementation, for that the new protocol version is introduced
* For partition based sync and upload progress everything is calculated in sync client code
* Provided estimate is expected to be a full replacement to any custom implementation based on existing transferred/transferable fields with subsequent deprecation and deletion of these callback parameters

Fixes #7450

## ☑️ ToDos
* [ ] 📝 Changelog update
* [x] 🚦 Tests (or not relevant)
* [x] C-API, if public C++ API changed.

papafe commented on src/realm/object-store/sync/sync_session.cpp:
```diff
@@ -1521,6 +1522,8 @@ SyncProgressNotifier::NotifierPackage::create_invocation(Progress const& current
 {
     uint64_t transferred = is_download ? current_progress.downloaded : current_progress.uploaded;
     uint64_t transferrable = is_download ? current_progress.downloadable : current_progress.uploadable;
+    double progress_estimate =
```
How come this value is 1.0 when `is_streaming` is true?

papafe replied:
I'm don't think this is the case here. I was actually trying our pre-existing tests (with PBS), and adding some conditions on the new progress estimate. And with streaming the estimate is always 1.0, while without it it gets a more reasonable value.

danieltabacaru commented on src/realm/sync/noinst/protocol_codec.hpp:
```diff
@@ -395,11 +406,26 @@ class ClientProtocol {
-        auto downloadable_bytes = msg.read_next<int64_t>();
+        DownloadInfoExtra info;
+        if (is_flx) {
+            info.query_version = msg.read_next<int64_t>();
```
we could update `peek_token_impl()` to handle doubles and the code below would be simplified.

kiburtse replied:
moved the implementation to HeaderLineParser

jbreams commented on src/realm/sync/client.cpp:
```diff
@@ -1102,6 +1106,28 @@ bool SessionImpl::is_steady_state_download_message(DownloadBatchState batch_stat
+void SessionImpl::notify_download_progress(const std::optional<uint64_t>& transient_changesets_size)
+{
+    if (m_state == State::Active) {
+        if (transient_changesets_size) {
```
After reading through the PR I still don't quite get what the difference between the transient changeset size and the downloaded bytes info we store in history is.

kiburtse replied:
i've simplified a bit the code around it so that it's clear where it is used - essentially these are bootstrap store pending bytes, it's not used anywhere else

jbreams commented on src/realm/sync/client.cpp:
```diff
@@ -1797,20 +1829,28 @@ void SessionWrapper::report_progress(bool only_if_new_uploadable_data)
+                         downloaded_bytes, downloadable_bytes, uploaded_bytes, uploadable_bytes, snapshot_version);
+
+    // FIXME take into account first initial reported value in this batch
+    double upload_estimate = uploadable_bytes > 0 ? (uploaded_bytes / double(uploadable_bytes)) : 1;
```
nit, this is not really an estimate - we know for sure this is how many uploadable bytes we have and how many bytes we have yet to upload.

kiburtse replied:
yeah.. i've also mentioned this multiple time during scoping, but it's not really important. Since bytes will go away at some point with pbs, then the only thing which is deemed important is the progress value itself. And it's somewhat obscure number, but it's the only useful thing for user facing progress. And it can't be precise, especially we use the same api in object-store for flx download direction.

kiburtse replied:
so, is it fine to resolve or would you like to weight in on changing it somehow from the original scope?

jbreams commented on src/realm/sync/client.cpp:
```diff
@@ -1797,20 +1829,28 @@ void SessionWrapper::report_progress(bool only_if_new_uploadable_data)
+        if (m_transient_downloaded_bytes)
+            downloaded_bytes += m_transient_downloaded_bytes.value();
+
+        // FIXME add some estimate to this
```
I think for FLX we can't estimate this and the download estimate is kinda what you get in lieu of downloadable bytes.

kiburtse replied:
yes, but i've added at least for now some logic to it, until we fully deprecate pbs and remove this api. That's what was desired from sdks which will still try non breaking release first.

jbreams reviewed (commented):
Here are some initial comments. I don't quite get what the transient downloaded bytes tracking does or how it's different from the downloadable bytes we already track in history.

nirinchev commented on src/realm/sync/client.cpp:
```diff
@@ -1102,6 +1106,28 @@ bool SessionImpl::is_steady_state_download_message(DownloadBatchState batch_stat
+    if (m_state == State::Active) {
+        if (transient_changesets_size) {
+            m_wrapper.m_transient_downloaded_bytes =
+                m_wrapper.m_transient_downloaded_bytes.value_or(0) + transient_changesets_size.value();
```
`value()` seems to be unavailable on iOS 11, which is the minimum version supported by the Swift SDK. It's a similar issue for lines 1844 and 1846.

danieltabacaru replied:
yeah. just dereference it

kiburtse replied:
done

kiburtse commented on src/realm/sync/client.cpp:
```diff
@@ -1102,6 +1106,28 @@ bool SessionImpl::is_steady_state_download_message(DownloadBatchState batch_stat
+
+void SessionImpl::notify_download_progress(const std::optional<uint64_t>& transient_changesets_size)
+{
+    if (m_state == State::Active) {
```
done

kiburtse commented on src/realm/sync/noinst/client_impl_base.hpp:
```diff
@@ -408,6 +407,7 @@ class ClientImpl::Connection {
     using SSLVerifyCallback = SyncConfig::SSLVerifyCallback;
     using ProxyConfig = SyncConfig::ProxyConfig;
     using ReconnectInfo = ClientImpl::ReconnectInfo;
+    using DownloadSyncProgressInfo = ClientProtocol::DownloadInfoExtra;
```
simplified with DownloadMessage struct

kiburtse commented on src/realm/sync/client.cpp:
```diff
@@ -1779,6 +1807,10 @@ void SessionWrapper::report_progress(bool only_if_new_uploadable_data)
         return;
 
+    // Ignore progress messages from before we first receive a DOWNLOAD message
+    if (!m_reliable_download_progress)
```
this was just a refactoring moving this exact check from object-store callback. I've separated changes under separate commit here, so you may see that it doesn't change the observed behaviour on object-store side. There was also fixme in this function about this.

kiburtse replied:
how would you tell if there is progress on upload then?

danieltabacaru replied:
there are two scenarios where we can report upload progress: when commiting (this decreases the progress estimate) and when receiving download messages (this increases the progress estimate)

danieltabacaru replied:
it could be I don't fully get the question, but we can report upload progress without needing any download message (keep reporting the committed data)

danieltabacaru replied:
uploaded is changed when we receive download messages. The use case I'm thinking is that the user should still get progress messages even if they are offline (if they make changes of course)

danieltabacaru replied:
we could also say we only report progress when data is uploaded (irrespective to what is committed)

kiburtse replied:
we do both already, no? uploadable is updated and propagates to object store api on commit notification, and uploaded is updated when acknowledged by the server with DOWNLOAD message.

This is a long discussion already, i'm not sure what is being asked here. Probably, even more confusing the original name of this variable how it was: m_reliable_download_progress 

It's not exactly 'reliable' for anything, essentially it was just blocking progress reporting until session receives first ever DOWNLOAD message. Probably, to make sure what the state of sync progress from the server side?

danieltabacaru replied:
> essentially it was just blocking progress reporting until session receives first ever DOWNLOAD message.

This contradicts with progress being reported if the user is offline.

kiburtse replied:
yes, right, what i meant is that we invoke progress update because of uploadable changes on commit, but not until once again first download message is received. So this variable m_reliable_download_progress was just like initial guard to prevent any notifications. And once again i'm still not sure what was the idea behind this, i've just done refactoring moving it away from object store into sync impl, since tests in both places seemed to diverge. So what do you think essentially? Should we just remove it?

kiburtse replied:
ok.. i'm now very confused with this. I'll do what you ask, no problem, i'm not going to argue since looks like my understanding is a bit off here. That will require adapt the code once again in tests mostly. 

just one last time to educate me on this one, i seek to understand the sync impl a bit better with this:

```
// so the code was in sync impl before:
void SessionWrapper::report_progress(bool only_if_new_uploadable_data) {
    ...

    uint64_t progress_version = (m_reliable_download_progress ? 1 : 0);
    m_progress_handler(..., progress_version, ...);
}
// .. and in object-store: 
void SyncProgressNotifier::update(..., uint64_t download_version, ...) {
    // Ignore progress messages from before we first receive a DOWNLOAD message
    if (download_version == 0)
        return;

    ...
}
```

what this subtly breaking behaviour be if it is checked now at the beginning here:
```
void SessionWrapper::report_progress(bool only_if_new_uploadable_data) {
    if (!m_reliable_download_progress)
        return;
    ...
}
```

jbreams replied:
Let's talk about this tomorrow.

jbreams replied:
Talked offline and this change makes sense to me now.

kiburtse commented on src/realm/sync/noinst/protocol_codec.hpp:
```diff
@@ -395,11 +406,26 @@ class ClientProtocol {
+
+            auto sv = msg.read_next<std::string_view>();
+            try {
+                info.progress_estimate = std::stod(std::string(sv));
```
overload for double is not yet available in libc++, but svo buffer is plenty big - string with double isn't expected to be more than 6 chars

kiburtse commented:
sync tests need a minor fix to make ci happy, other then that object-store tests should pass and the pr is almost done

kiburtse commented:
> sync tests need a minor fix to make ci happy, other then that object-store tests should pass and the pr is almost done

fixed, ci should be green now.

@danieltabacaru @jbreams please, take a look. All commits here a logically separated, it could be easier to review them one by one. The only missing thing is final master merge and changelog. I'll do it after a few ci runs.

jbreams commented on src/realm/sync/client.cpp:
```diff
@@ -1692,12 +1750,22 @@ inline void SessionWrapper::finalize_before_actualization() noexcept
 
-inline void SessionWrapper::on_sync_progress()
+    if (!only_if_new_uploadable_data)
+        m_reliable_download_progress = true;
```
is this okay because we can't have any uploads until we've received at least one download? if so, can you add a comment explaining that?

jbreams replied:
Any updates on this?

kiburtse replied:
added the comment with a few details

jbreams commented on src/realm/sync/noinst/client_impl_base.cpp:
```diff
@@ -1601,21 +1597,31 @@ void Session::on_changesets_integrated(version_type client_version, const SyncPr
     bool upload_progressed = (progress.upload.client_version > m_progress.upload.client_version);
+    bool download_progressed =
+        progress.download.server_version > m_progress.download.server_version &&
+        progress.download.last_integrated_client_version == m_progress.download.last_integrated_client_version;
```
hmm. why this second check to see if the last_integrated_client_version hasn't progressed? i kinda thought this would be timing dependent on the server.

kiburtse replied:
Server version increases on every upload we make, and such DOWNLOAD message has no changesets in it.
Don't we hold off uploads until server signals that there are no more changesets for us to get?
Maybe i misunderstand this, but what is considered a progress in download?

jbreams replied:
Tldr, The only thing that advances progress is receiving a DOWNLOAD, but there are no guarantees about what will be in the download that advances your progress.

We hold off on uploads when the session first starts, but after we receive the first batch of downloads to get caught up uploads can happen at any time. The server also will respond to an upload with an empty download if there are no further changesets to send but progress advanced, but the server could also just have more changesets to send and you'll do an upload and receive a download with changesets from before the upload and a new progress cursor indicating the server has accepted/persisted your upload.

kiburtse replied:
Replaced it to account for that i believe. These checks with client versions were not really needed.

jbreams reviewed (commented):
Great progress, here are some comments/questions. We're almost there!

danieltabacaru commented on src/realm/sync/client.cpp:
```diff
@@ -1784,47 +1852,90 @@ void SessionWrapper::on_connection_state_changed(ConnectionState state,
+    if (m_download_estimate) {
+        download_estimate = *m_download_estimate;
+
+        // ... and transient bytes are only relevant with bootstrap store
```
we should then name it accordingly (i.e, bootstrap_downloaded_bytes), since this is a bit confusing

danieltabacaru commented on src/realm/sync/client.cpp:
```diff
@@ -1784,47 +1852,90 @@ void SessionWrapper::on_connection_state_changed(ConnectionState state,
+        if (m_transient_downloaded_bytes)
+            p.downloaded += *m_transient_downloaded_bytes;
+
+        // FIXME for flx with download estimate these bytest are not known
```
bytes

danieltabacaru commented on src/realm/sync/client.cpp:
```diff
@@ -1784,47 +1852,90 @@ void SessionWrapper::on_connection_state_changed(ConnectionState state,
+
+    if ((!is_completed || is_download) && p.uploaded != p.uploadable) {
+        if (p.uploadable > p.final_uploaded)
+            upload_estimate = (p.uploaded - p.final_uploaded) / double(p.uploadable - p.final_uploaded);
```
why do the diff? shouldn't the estimate be `p.uploaded/p.uploadable`? IIUC, final bytes only make sense for download (where the bootstrap is added)

kiburtse replied:
This values are always simply increasing, so if you add 1mb of data you will get values from 0 to 1048576.
Then if you add one more 1mb reported values are going to be 1048576 to 2097152, and so on.
So the progress reported will be always shorter and shorter range first from 0.0 to 1.0, then from 0.5 to 1.0, then 0.66 to 1.0. How does it make sense? That is also not how server sends its progress for download. It always restarts after completion.

danieltabacaru replied:
I think the reason why the progress always increases for download is because the server (cc @rohitn212) associates the remaining changes with a server version, and then it keeps it at 1.0 past that unless there is a change of query. For upload things are a bit different (since we know exactly the changes we need to upload and also there is no bootstrap), so I think it's fine for progress to increase and decrease as changes are committed and acknowledged by the server. Alternatively, we can set the last changeset as the one the progress will report 1.0 for, then decrease and increase towards 1.0 until again the last changeset is uploaded and so on. I am not sure what strategy we follow with these changes.

rohitn212 replied:
Yes, on the server side, we ensure that the progress estimate remains non-increasing. We also only send the full 100% (1.0) value for the final download changeset.

rohitn212 replied:
Oops, yes. Sorry, that was a typo. 
The COQ wouldn't be processed until after whatever triggered the initial history scan (ex BindQuery). But yes, once the initial history scan is complete and the client has caught up after reconnecting then the progress will end at 1.0.
The COQ would be a separate instance so the progress starting from 0 doesn't break the non-decreasing rule.

danieltabacaru replied:
interesting. i was not aware a coq is postponed in this case.

rohitn212 replied:
If the client reconnected with just a Bind query then it wouldn't trigger the initial history scan.
@sudssm can you confirm whether what I said above is correct?

kiburtse replied:
@danieltabacaru what's your suggestion for this approach? I've tweaked testing code to check a few different scenarios. There is 'dump' function for progress values reported which is invoked on test failure, but you may simply call in on every 'clear(progress)' call to see which values were reported for every type of progress callback in upload/download pbs/flx. With this logic i saw most consistent behaviour. Do you think it is still unreasonable to proceed this way? Or is it fine to resolve?

sudssm replied:
@rohitn212 that's correct. but this code looks like its upload progress, which we havent made any changes to?
for download progress, the server will always send monotonically increasing values between 0 and 1, for each bootsrap or COQ. COQs are processed in serial, and we will not start one until the previous (bootstrap or COQ) is complete

jbreams replied:
I think we'll have to chat in person about this.

jbreams replied:
I wrote a test to see what's going on here and I think I've figured it out.
```
TEST_CASE("flx: progress debugging", "") {
    FLXSyncTestHarness harness("flx_bootstrap_reconnect", {g_large_array_schema, {"queryable_int_field"}});
    SyncTestFile uploader_config(harness.app()->current_user(), harness.schema(), SyncConfig::FLXSyncEnabled{});

    auto uploader_realm = Realm::get_shared_realm(uploader_config);
    {
        auto mut_subs = uploader_realm->get_latest_subscription_set().make_mutable_copy();
        auto table = uploader_realm->read_group().get_table("class_TopLevel");
        mut_subs.insert_or_assign(Query(table));
        mut_subs.commit().get_state_change_notification(sync::SubscriptionSet::State::Complete).get();
    };

    enum State { Start, FilledInitialData, InNotifier, FilledSecondData };
    TestingStateMachine<State> state(Start);

    auto logger = util::Logger::get_default_logger();
    auto fill_realm = [generation = 1, logger](SharedRealm realm) mutable {
        CppContext c(realm);
        for (int i = 0; i < 5; ++i) {
            realm->begin_transaction();
            auto obj =
                Object::create(c, realm, "TopLevel",
                               std::any(AnyDict{{"_id", ObjectId::gen()},
                                                {"list_of_strings", AnyVector{}},
                                                {"queryable_int_field", static_cast<int64_t>(i + generation)}}));
            List str_list(obj, realm->schema().find("TopLevel")->property_for_name("list_of_strings"));
            for (int j = 0; j < 1024; ++j) {
                str_list.add(c, std::any(std::string(1024, 'a' + (j % 26))));
            }

            realm->commit_transaction();
            logger->info("Committed transaction for object %1", i + generation);
        }
        generation += 100;
    };

    auto progress_nofitier = [&](uint64_t transferred_bytes, uint64_t transferrable_bytes, double) {
        bool should_wait = false;
        state.transition_with([&](State cur_state) -> std::optional<State>{
            if (cur_state == FilledInitialData && transferred_bytes > 0 && transferrable_bytes != transferred_bytes) {
                should_wait = true;
                return InNotifier;
            }
            return std::nullopt;
        });
        if (should_wait) {
            state.wait_for(FilledSecondData);
        }
    };
    uploader_realm->sync_session()->register_progress_notifier(std::move(progress_nofitier),
                                                               SyncSession::ProgressDirection::upload, true);
    fill_realm(uploader_realm);
    state.transition_with([&](State cur_state) {
        REQUIRE(cur_state == Start);
        return FilledInitialData;
    });
    state.wait_for(InNotifier);
    fill_realm(uploader_realm);
    state.transition_with([&](State cur_state) {
        REQUIRE(cur_state == InNotifier);
        return FilledSecondData;
    });
    wait_for_upload(*uploader_realm);
}
```

I think the logic here is that if you don't do this difference calculation between uploadable/uploaded/final_uploaded, when you add a bunch of stuff to upload, the progress estimate doesn't go back to zero, it goes back to some percentage non-zero percentage. This calculation lets you have a clean progression from zero to 1.0 even if you then add more data to upload before it reaches 1.0, and then once you reach 1.0 the next batch of uploads will restart the estimate at zero.

Can we just add a comment to that effect?

danieltabacaru replied:
I ran the test and the progress goes back to zero only after the upload is complete, not if more data is added before progress reaches 1.0.

@jbreams Not sure what you mean by:
> This calculation lets you have a clean progression from zero to 1.0 even if you then add more data to upload before it reaches 1.0

Is it that it always increases? Because that's not the case (try changing transferred_bytes > 0 into transferred_bytes > 78). The progress goes from 0.2 to 0.1 (the same with my proposed calculation).

I think this is in line with what I mentioned here though:
> Alternatively, we can set the last changeset in history as the one the progress will report 1.0 for, then decrease and increase towards 1.0 until again the last changeset is uploaded and so on.

All in all, this has the nice property that it behaves (more or less) similarly with download progress.

jbreams replied:
Sorry, I meant that it goes from zero to 1.0 until everything is uploaded, but when everything is uploaded and you add new data it resets to zero, like this 0 -> 1.0 -> 0.1 -> 0.0 -> 0.1 ... -> 0.4 -> new data added -> 0.3 ... -> 1. Otherwise it goes from 0 -> 1.0 -> 0.0 -> 0.1 ... -> 0.4 -> new data added -> 0.3 ... -> 1.

danieltabacaru commented on src/realm/sync/client.cpp:
```diff
@@ -1784,47 +1852,90 @@ void SessionWrapper::on_connection_state_changed(ConnectionState state,
+        p.downloadable = p.downloaded;
+        if (0.01 <= download_estimate && download_estimate <= 0.99)
+            if (p.downloaded > p.final_downloaded)
+                p.downloadable = p.final_downloaded + (p.downloaded - p.final_downloaded) / download_estimate;
```
similar to upload, shouldn't this be `p.downloadable = p.final_downloaded / download_estimate`?

danieltabacaru commented on src/realm/sync/client.cpp:
```diff
@@ -1784,47 +1852,90 @@ void SessionWrapper::on_connection_state_changed(ConnectionState state,
+}
 
-void SessionWrapper::report_progress(bool only_if_new_uploadable_data)
+void SessionWrapper::report_progress(bool is_download, bool is_completed, bool only_if_new_uploadable_data)
```
I find the logic in this function hard to follow (with a lot of checks). I will put some thought into a cleaner approach.

kiburtse replied:
i also want to make it cleaner with #7451 Reusing the same method for clear intent in this change was simpler initially, but became pretty ugly as you've noticed. I've almost finished the next change splitting this function for download and upload, so hopefully it'll get better.

danieltabacaru reviewed (commented):
Left some comments after my first pass (I haven't reviewed the tests yet). There are some assumptions (such as using the upload completion - don't know why it's even needed) which may not hold, so there may be edge cases not covered. I will put some thought into how some of the code could be improved.

kiburtse commented on src/realm/sync/client.cpp:
```diff
@@ -1784,47 +1852,90 @@ void SessionWrapper::on_connection_state_changed(ConnectionState state,
+    if (!m_reliable_download_progress)
+        return;
+
+    ReportedProgress p = m_reported_progress;
```
ReportedProgress has also snapshot version in it which is used below for downloads to decide if progress has changed. I should probably try to make it cleaner.

jbreams commented on src/realm/sync/client.cpp:
```diff
@@ -1600,8 +1656,10 @@ void SessionWrapper::actualize(ServerEndpoint endpoint)
 
-    if (!m_client_reset_config)
-        report_progress(); // Throws
+    if (!m_client_reset_config && m_reliable_download_progress) {
```
So looking at the origins of this line. It got added in this PR https://github.com/realm/realm-sync/pull/2779, and I think it was to resolve this issue https://github.com/realm/realm-sync/issues/2859. I think the idea is that we want to report that the session has been activated and has received zero progress so that apps can start a spinner/progress bar in their UI. Why do we only want to call report_progress() here if we have reliable download progress?

kiburtse replied:
I'm not sure that is the case here since essentially to understand the behaviour of this one would need to look at both sides: sync client impl and object store notifier. Looks like it's changed a few time (or fixed) but it's hard to tell what was the intent really. I think https://github.com/realm/realm-sync/issues/2859 is different issue: it essentially asks to report progress the moment client is about to start downloading the changeset every time. I don't think it makes sense to do it on this level in client.cpp, but anyway the issue is still open. The suggested change fixing this was probably irrelevant.
Toggle to m_reliable_download_progress == true happens only once on connect in current implementation, so my guess is that it was there to report initial sync status on connect with changesets to upload. I don't think it is useful in any other way.
I believe I've changed it now to be explicit about how it behaved before. I don't see anything breaking with this or changing observed behaviour. What else should be done here? What do you think?

jbreams commented on src/realm/sync/noinst/protocol_codec.hpp:
```diff
@@ -87,7 +87,7 @@ class HeaderLineParser {
     {
         // We currently only support numeric, string, and boolean values in header lines.
-        static_assert(std::is_integral_v<T> || is_any_v<T, std::string_view, std::string>);
+        static_assert(std::is_integral_v<T> || is_any_v<T, std::string_view, std::string, double>);
```
to be fair, `std::is_floating_point_v` would cover a lot of types besides double, and double is the only type we actually handle here.

danieltabacaru replied:
right. not a strong opinion, but why not properly support all floating point types?

jbreams commented on src/realm/sync/noinst/protocol_codec.hpp:
```diff
@@ -376,10 +394,22 @@ class ClientProtocol {
         }
     }
 
+    struct DownloadMessage {
```
Agree. I like having the DownloadMessage type and would be fine folding received_changesets into this.

jbreams reviewed (commented):
Just some quick responses. Looks like there are still unaddressed comments, so let me know when this is ready for review again.

danieltabacaru commented:
Maybe we should print the progress estimate with a fixed number of decimals:
`Realm.Sync.Client.Session - Connection[1]: Session[1]: Progress handler called, downloaded = 75, downloadable = 75, estimate = 1, uploaded = 78, uploadable = 5365424, estimate = 1.45375e-05, snapshot version = 17`

kiburtse commented:
> Maybe we should print the progress estimate with a fixed number of decimals: `Realm.Sync.Client.Session - Connection[1]: Session[1]: Progress handler called, downloaded = 75, downloadable = 75, estimate = 1, uploaded = 78, uploadable = 5365424, estimate = 1.45375e-05, snapshot version = 17`
> 
> (from the server spec: `Download Progress Percentage will be a percentage representation of how much progress has been made so far, of type float, ranging from 0 to 1 with up to 4 points of precision.`)
> 
> There also seems to be a division problem: `Realm.Sync.Client.Session - Connection[1]: Session[2]: Progress handler called, downloaded = 5365700, downloadable = 5365700, estimate = 0.9999, uploaded = 78, uploadable = 78, estimate = 1, snapshot version = 18` The estimate should be 1.0 in this case.

Fixed log entries with the precision agreed on the scope to 4 digits

0.9999 was actually the corner case bug hidden by the workaround for the other issue (there is still fixme about this #7452)

@danieltabacaru @jbreams i believe i've addressed all opened threads, please, take a look again.

danieltabacaru commented on src/realm/sync/noinst/client_impl_base.cpp:
```diff
@@ -2759,4 +2786,5 @@ void Session::check_for_download_completion()
         ensure_enlisted_to_send(); // Throws
     }
     on_download_completion(); // Throws
+    return;
```
?

danieltabacaru commented on src/realm/sync/noinst/client_impl_base.cpp:
```diff
@@ -2738,6 +2764,7 @@ void Session::check_for_upload_completion()
 
     m_upload_completion_notification_requested = false;
     on_upload_completion(); // Throws
+    return;
```
?

danieltabacaru commented on src/realm/sync/client.cpp:
```diff
@@ -1784,47 +1853,107 @@ void SessionWrapper::on_connection_state_changed(ConnectionState state,
+            p.downloaded += *m_bootstrap_store_bytes;
+
+        // FIXME for flx with download estimate these bytes are not known
+        // provide some sensible value for non-streaming version of object-store callbackas
```
callbacks

jbreams commented on test/object-store/sync/session/progress_notifications.cpp:
```diff
@@ -672,8 +607,304 @@ TEST_CASE("progress notification", "[sync][session][progress]") {
+        wait_for_upload(*realm);
+        wait_for_download(*realm);
+        realm->sync_session()->pause();
+        realm->refresh();
```
ping

jbreams replied:
wait_for_advance waits for the Realm to be advanced to the latest snapshot, refresh() updates the realm to whatever the latest snapshot is at the time you call it. basically wait_for_advance() exercises the notifier machinery to auto-advance your realm to the latest thing and refresh doesn't. you do need to actually have a read transaction to use it tho. i think you could stick a `realm->read_group()` before calling it.

jbreams replied:
Talked about this offline, and decided calling refresh was the right thing.

danieltabacaru commented on src/realm/sync/noinst/client_impl_base.cpp:
```diff
@@ -1598,25 +1593,35 @@ void Session::on_integration_failure(const IntegrationException& error)
+    bool resume_upload = do_recognize_sync_version(client_version); // Allows upload process to resume
+
+    // notify also when final DOWNLOAD received with no changesets
+    bool download_progressed = changesets_integrated || (!upload_progressed && resume_upload);
```
Thinking of it a bit more, isn't `bool download_progressed = changesets_integrated` enough? For download, we should have something to report only when we integrate changesets. It's true that a download has implications on upload too (but that's reported separately and we handle it above)

danieltabacaru replied:
I looked at the original code, and to have the same behavior (+ bootstrap bytes) you'd need to always invoke `notify_download_progress` here (on_sync_progress was essentially moved to on_changesets_integrated). I am wondering though, don't we sometimes get the same progress twice (if both notify_download_progress and notify_upload_progress are invoked in this function)?

I'll take a look at those tests.

kiburtse replied:
Interesting, so that may explain why i haven't noticed any issues in object-store tests. I haven't found any assertion or a check in sync code, so i've just assumed here that if explicit check for upload_progressed yields false, then it must be download related update. So it might have been corner case in test sync server implementation - DOWNLOAD message without changesets and change in client and server versions in it.

Anyway, duplicated values are essentially already the problem (i've noted it here #7451), and i'd like to ultimately solve it. Here i've made explicit split for upload/download notify, and the only situation where you'll get the same progress twice is when DOWNLOAD has changesets and acknowledged upload. If you think it should be changed here, then the fix is straightforward: omit report_progress for upload and call here notify_download_progress unconditionally. At least until #7451 is fully solved. I've refactored it already to work for object-store tests, but need still some work on sync tests, because they are a bit weird.

jbreams replied:
I think splitting on_sync_progress into on_download_progress/on_upload_progress has made a bunch of these checks more confusing. Before it was unclear if you didn't know that downloads are the only thing that advances progress (but at least you only had to look in one place to figure that out), and now it's confusing all the time (and you have to check in two places). I don't want to keep arguing about this though.

danieltabacaru replied:
what is the conclusion here?


> about notify_download_progress: it is essentially the replacement for old on_sync_progress and it was always called unconditionally along on_changesets_integrated.

It seems we should always call `notify_download_progress`, right?

danieltabacaru commented on src/realm/sync/client.cpp:
```diff
@@ -266,7 +279,8 @@ class SessionWrapper final : public util::AtomicRefCountBase, DB::CommitListener
 
-    void report_progress(bool only_if_new_uploadable_data = false);
+    void init_progress_handler();
+    void report_progress(bool is_download, bool only_if_new_uploadable_data = false);
```
could add a comment that `only_if_new_uploadable_data` can be true only if `is_download` is false

kiburtse replied:
added the comment and the check

danieltabacaru commented on src/realm/sync/client.cpp:
```diff
@@ -868,8 +881,13 @@ bool SessionImpl::process_flx_bootstrap_message(const SyncProgress& progress, Do
         return true;
     }
+    else {
+        // FIXME this shouldn't be needed anymore when progress is reported separately for every direction
```
is this comment correct? I guess we need to reset it somewhere (I would do it when the progress gets to 1.0)

kiburtse replied:
yes, this is correct. It's possible to remove this variable completely and simply pass it through notify_download_progress as a temporary but it's really only possible with solving #7451. I've rephrased this fixme to better communicate the intent here.
Also, this is essentially the place in code where it's known that we don't get more changesets for bootstrap store, so i'd like to keep it here to not accidentally report doubled bytes.

danieltabacaru commented on src/realm/sync/client.cpp:
```diff
@@ -1600,8 +1648,10 @@ void SessionWrapper::actualize(ServerEndpoint endpoint)
-        report_progress(); // Throws
+    if (!m_client_reset_config) {
+        // FIXME looks like noop, could be removed without an issue?
+        on_upload_progress(/* only_if_new_uploadable_data = */ true); // Throws
```
I would keep it for now (so we definitely have the same behavior) and remove the comment

danieltabacaru commented on test/tsan.suppress:
```diff
@@ -19,3 +19,8 @@ deadlock:realm::sync::MigrationStore::create_subscriptions
 race:adjtime
+
+# libuv scheduler is only supposed to be used on main thread but constructed
+# on the sync thread for asyn open task, should be harmless to suppress,
```
async

danieltabacaru commented on test/test_sync.cpp:
```diff
@@ -2986,65 +2986,52 @@ TEST(Sync_UploadDownloadProgress_1)
-            if (handler_entry == 0) {
+        std::unique_lock<std::mutex> lock(mutex);
+        session.set_progress_handler(progress_handler);
+        session.set_connection_state_change_listener([&](ConnectionState state, util::Optional<ErrorInfo>) {
```
why do we need the connection state to change? is there any progress reported before that we want to skip?

kiburtse replied:
this was the part of `m_reliable_download_progress` variable check refactoring. Before it was essentially checking the condition for `progress_version == 0`, and this funcionaly the same as to check that callback wasn't called after we connected but not yet received download message.

So you can see below that the snapshot version is still the same after the first commit. And that is exactly how object-store api was receiving it before refactoring.

danieltabacaru commented on test/test_sync.cpp:
```diff
@@ -3678,27 +3624,35 @@ TEST(Sync_UploadDownloadProgress_6)
         CHECK_EQUAL(uploadable_bytes, 0);
-        CHECK_EQUAL(progress_version, 0);
-        CHECK_EQUAL(snapshot_version, 1);
+        CHECK_EQUAL(snapshot_version, 3);
```
why did this change? the setup seems the same.

kiburtse replied:
first progress call with `progress_version == 1` was with the `snapshot_version == 3`, so it's functionally the same checks adapted on `m_reliable_download_progress` refactoring

danieltabacaru commented on test/object-store/sync/session/progress_notifications.cpp:
```diff
@@ -672,8 +607,383 @@ TEST_CASE("progress notification", "[sync][session][progress]") {
+        : harness(app_id)
+    {
+        table_name = (*harness.schema().begin()).name;
+        harness.do_with_new_user([](std::shared_ptr<SyncUser>&&) {});
```
I don't think this is needed. But if you want an user, `create_user_and_log_in` would suffice.

danieltabacaru commented on test/object-store/sync/session/progress_notifications.cpp:
```diff
@@ -672,8 +607,383 @@ TEST_CASE("progress notification", "[sync][session][progress]") {
+ *
+ * Test emulates adding a few changes into one realm, verifies that the progress is reported until upload completion,
+ * and then checks how this exact changes are downloaded into the second realm file (check with bootstrap store for
+ * flx). Then without realm instances reset it is the other way around: test adds more changes to the second realm and
```
nit: this sentence can be rephrased
